include/cinder/linux/GstPlayer.h

```diff
@@ -16,212 +15,226 @@
+	
+	void 			load( const std::string& path );
+	bool 			newVideo() const;
+	
+	void 			play();
+	void 			stop();
+	
+	int 			width() const;
+	int 			height() const;
+
+	bool 			isPrerolled() const;
+	bool 			isPaused() const;
+	bool 			isPlayable() const;
+	bool 			isLoaded() const;
+	bool 			isBuffering() const;
+	bool 			isLiveSource() const;
+	int  			stride() const;
+	void 			setLoop( bool loop = true, bool palindrome = false );
+	bool 			setRate( float rate );
+	bool 			stepForward();
+	
+	float 			getRate() const;
+	
+	bool 			hasNewFrame() const;
+	
+	GstVideoFormat		format() const;
+
+	gint64 			getPositionNanos();
+	float 			getPositionSeconds();
+	gint64 			getDurationNanos();
+	float 			getDurationSeconds();
+	float                   getFramerate() const;
+	float			getPixelAspectRatio() const;
+	int 		        getNumFrames();
+	
+	bool                    hasAudio() const;
+	bool                    hasVisuals() const;
+
+	void 			setVolume( float targetVolume );
+	float 			getVolume();
+	
+	bool 			isDone() const;
+	
+	GstElement* 		getPipeline();
+	
+	void 			seekToTime( float seconds );
+        void                    seekToFrame( int frame );
+	
+	bool 			isStream() const;
+
+	ci::gl::Texture2dRef	getVideoTexture();
+
+private:		
+	bool 			initializeGStreamer();
+
+	void                    constructPipeline();
+
+	void 			startGMainLoopThread();
+	void 			startGMainLoop( GMainLoop* loop );
+	
+	static void 		onGstEos( GstAppSink* sink, gpointer userData );
+	static GstFlowReturn 	onGstSample( GstAppSink* sink, gpointer userData );
+	static GstFlowReturn	onGstPreroll( GstAppSink* sink, gpointer userData );
+	void 			processNewSample( GstSample* sample );
+	void 			getVideoInfo( const GstVideoInfo& videoInfo );
+	
+	bool 		        setPipelineState( GstState targetState );
+	bool 			checkStateChange( GstStateChangeReturn stateChangeResult );
+	GstStateChangeReturn 	getStateChange();
+
+	GstState 		getCurrentState();
+	GstState 		getPendingState();
+
+	bool			sendSeekEvent( gint64 seekTime );
+	
+	void			addBusWatch( GstElement* pipeline );
+
+	void 			resetCustomPipeline();
+	void 			resetPipeline();
+	void 			resetBus();
+	void 			cleanup();
+	void 			resetVideoBuffers();
+	void 			resetGLBuffers();
+	void 			resetSystemMemoryBuffers();
+
+	void 			createTextureFromMemory();
+	void 			createTextureFromID();
+
+private:
+	GMainLoop* 		mGMainLoop; // Needed for message activation since we are not using signals.
+	GstBus* 		mGstBus; // Delivers the messages.
+	int  			mBusId; // Save the id of the bus for releasing when not needed.
+	std::thread	 	mGMainLoopThread; // Seperate thread for GMainLoop.
+	
+	std::mutex 		mMutex; // Protect  since the appsink callbacks are executed from the streaming thread internally from GStreamer.
+
+	bool 			mUsingCustomPipeline;
+	GstData 		mGstData; // Data that describe the current state of the pipeline.
+	
+	ci::gl::Texture2dRef	mVideoTexture;
+
+	unsigned char* 	    	mFrontVBuffer = nullptr;

```

Comment:
Suggestion: instead of using raw pointers here, I like using vector<char> and then passing off its .data(). It avoids having to clean up and then it is very easy and legible to resize as necessary.